pkg/controllermanager/controller/controllerregistration/seed_control.go

```diff
@@ -25,21 +25,36 @@ import (
 	"github.com/gardener/gardener/pkg/controllerutils"
 	"github.com/gardener/gardener/pkg/logger"
 
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/client-go/tools/cache"
 )
 
-func (c *Controller) seedAdd(obj interface{}) {
+func (c *Controller) seedAdd(obj interface{}, addToControllerRegistrationQueue bool) {
```

Comment:
Can you elaborate why we need any controller registration handling if we already have dedicated `defaultControllerRegistrationSeedControl` controller watching seed events ([ref](https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go))?

Reply:
Background:
When a dnsProvider is specified in the Seed resource we want the dns extension to be installed in the Seed (if not present yet). In order to detect whether a dnsProvider is needed we need to watch the Seed resource (spec.dns.provider).
In the code if a Seed resource is [updated](https://github.com/gardener/gardener/blob/52a23a2d77ca8dda65fa592752b91becfda8d79e/pkg/controllermanager/controller/controllerregistration/seed_control.go#L46) we check whether a the dnsProvider changed. If yes we add the Seed to the [controller-registration-seed-queue](https://github.com/gardener/gardener/blob/52a23a2d77ca8dda65fa592752b91becfda8d79e/pkg/controllermanager/controller/controllerregistration/seed_control.go#L42). In this queue it will check if a dns-extension is needed [here](https://github.com/gardener/gardener/pull/3131/files#diff-0d8996ff3546653f546482d70a198c7c4da27e0102f684a87d5e982df34896d4R163). This is also the Controller you mentioning above I assume.

@timuthy and me had a discussion whether we can implement it similar to how we watch backupBuckets [here](https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration.go#L114) and watch for when a dnsProvider gets added. But that won't work as the backupBucket and the dnsProvider are different in nature. The backupbucket resource watched is gardencorev1beta1.BackupBucket which holds a reference to the Seed. There is no equivalent resource for the dnsProvider. Therefore, we can not watch the dnsProvider resource and need to watch the Seed resource as it is implemented right now. So if you have no further objections I would leave the implementation as it is right now.